### PR TITLE
Upgrade web-vitals to 3.3.1 and switch to attribution build

### LINKED ISSUE #77: Log CWV events to the console
**Is your feature request related to a problem? Please describe.**
The CWV values reported by the extension are useful snapshots of the UX, but they don't necessarily have any diagnostic information for developers to understand what could have caused adverse experiences.

**Describe the solution you'd like**
I'd like to see the extension utilize console logging more to leave an audit trail to assist developers with debugging CWV issues. For example, log the times at which CLS changes and if possible attribute each shift to the culprit element. Or for LCP, log the elements that are considered the largest and the times at which they are painted. For FID, it'd be helpful to know the timestamp of the first input and what was clicked.

The extension can be a first class developer tool for not only understanding how pages are experienced but also debugging how to make them better. Stretch goal: maybe like a VisBug for performance?

### LINKED ISSUE #110: Upgrade to the latest web-vitals.js attribution build

### LINKED ISSUE #112: Add measures and logging for LCP diagnostics
The [Optimize LCP guide](https://web.dev/optimize-lcp/#monitor-lcp-breakdown-in-javascript) describes how to instrument DevTools with additional LCP diagnostic info. It neatly logs each diagnostic in a table and uses `performance.measure()` to show them in the Performance panel.

Figure out whether upgrading to the attribution build in #110 is sufficient for logging the diagnostics, or whether we should add any extra table formatting like the example in the guide.

The `measure` feature should be gated behind the same User Timings option used by INP.

### PR DESCRIPTION
**Pull request by tunetheweb**
Progress on #110 
Fixes #112
Fixes #77 
Progress on #86 

This PR makes the following changes:
- Upgrades web-vitals to the latest version (3.3.1)
- Switches to the attribution build
- Fixes a bug where userTiming also needed console.log selected
- Adds a console table option to display the breakdowns in console tables (see screenshot below)
- Adds LCP breakdowns to both Performance.measure and console
- Adds CLS, FID, INP breakdowns to console.log
- Consistently uses '[Web Vitals Extension]' in console logging now

image

Here's what that looks like:

image

### CODE REVIEW

File: src/browser_action/vitals.js
Comment (mmocny) [inline src/browser_action/vitals.js:21]: We used the term `User Timing` specifically to mean performance.mark() and measure() since the API for that is called User Timing: https://developer.mozilla.org/en-US/docs/Web/API/Performance_API/User_timing ... its a bit odd to say UserTimingConsole.


I see that you want to use a shared name since these both use shared data.

How about something like:

- enableAttributionToConsole
- enableAttributionToUserTiming

Another alternative is:

- Enable Console [checkbox]
- Enable User Timings [checkbox]
- Durations only, or, full attribution [radio button]

Perhaps we would even have checkbox for each of the 4 metrics so you can disable specific ones in case its noisy...

**Review by mmocny — COMMENTED** [review]
I like the changes (especially more sharing with web-vitals lib, nice), but I think we could fiddle with the settings a bit to change the permutations.

See comments inline.

File: src/browser_action/vitals.js
Comment (tunetheweb) [reply src/browser_action/vitals.js]: I renamed it to `consoleTables`/`enableConsoleTables`

Options table looks like this now:

image

File: src/browser_action/vitals.js
Comment (mmocny) [inline src/browser_action/vitals.js:238]: Since we're reporting "reportAllChanges" for these metrics, that means we'll log and create user timing for all metric candidates.  That's mostly a good thing, except maybe for LCP...

How does this look in DevTools when you load a page with multiple candidates reported over time?  Wont the timings overlap (TTFB etc) since all are reporting from 0?

Is there any way to know which when LCP is finalized with web-vitals.js?  We could console.log() all but user timings only for final?

Comment (tunetheweb) [reply src/browser_action/vitals.js:238]: We're not using `reportAllChanges`. And I question the value of doing that since it's really "reportAllChangesThatAreBiggerThanTheLast", so still only partially useful to report all layout shifts (CLS) or interactions (for INP).

The only thing that it would be good for, would be reporting LCP right away (i.e. before a click) and same for CLS and INP (the extension currently requires you to tab away and back before it reports these). But if we did ever introduce that, then yes we'd need to rethink how we did this. But, as I say, I question it's value.

Comment (tunetheweb) [reply src/browser_action/vitals.js:238]: Fixed in https://github.com/GoogleChrome/web-vitals-extension/pull/115/commits/991ff048d6947e9c40d000880abbd1761566a424

I think reporting multiple LCPs in this case is find. Not ideal but fine. It's already what happens with the existing console logging.

**Review by mmocny — APPROVED** [review]
Cool, LTGTM.

Several optional follow-ups but no blockers.

File: src/browser_action/vitals.js
Comment (rviscomi) [inline src/browser_action/vitals.js:251]: Getting this error:

> Uncaught (in promise) TypeError: Failed to execute 'measure' on 'Performance': Failed to read the 'start' property from 'PerformanceMeasureOptions': The provided double value is non-finite.
 
Seems like the TTFB value isn't part of the `lcpEntry` object: `metric.attribution.timeToFirstByte`. I think this applies to all of the LCP diagnostic measures.

**Review by rviscomi — CHANGES_REQUESTED** [review]

File: src/browser_action/vitals.js
Comment (tunetheweb) [reply src/browser_action/vitals.js:251]: I can't repeat this, and TTFB is part of metric.attribution.timeToFirstByte as shown with the console.log entry:

image

I recall having a similar error on an older version which I fixed but struggling to remember in which commit. Can you check you're looking at the latest version on this branch?

Comment (rviscomi) [reply src/browser_action/vitals.js:251]: Right, the bug is that the correct path is `metric.attribution.timeToFirstByte` but you're attempting to access it from `metric.attribution.lcpEntry.timeToFirstByte`.

This is the first time I've pulled from your branch, so it should be the latest version. Are you able to see the LCP timings in the Perf panel?

Comment (tunetheweb) [reply src/browser_action/vitals.js]: Ah see it now. Was looking at console.log. Fixed now.

File: src/browser_action/vitals.js
Comment (rviscomi) [inline src/browser_action/vitals.js:299]: I'd suggest reverting the CLS and INP tables. Don't think they're as useful as the LCP use case and not in scope for #112 anyway.

File: src/browser_action/vitals.js
Comment (rviscomi) [inline src/browser_action/vitals.js:312]: Aside: any reason these INP sub-parts aren't a standard part of `web-vitals` attribution?

**Review by rviscomi — CHANGES_REQUESTED** [review]

File: src/browser_action/vitals.js
Comment (tunetheweb) [reply src/browser_action/vitals.js]: > Convert stringy numbers to numbers (and remove unnecessary sub-ms precision)

Done (though using `Math.round(x,0)` rather than converting to String with toFixed and then back to number)

> Write full metric names

Done, though used Sentence case as per Optimize LCP guide and usual convention.

> Remove LCP

Disagree. I see this table as a summary presentation of the Performance Entry (which also means you can turn the PerformanceEntry option off completely). It's also helpful to give more context (the element name) when there are multiple LCP entries surfaced (which is an annoying thing that happens since we use `reportAllChanges` to ensure the extension UI is updated as soon as possible):

image

> Refer to metrics as sub-parts [rather than breakdown]

That would probably make sense if we were not including LCP itself which, as per above, I'd like to keep.

It's also referred to as "Breakdown" in the Optimize LCP guide: https://web.dev/optimize-lcp/#lcp-breakdown

File: src/browser_action/vitals.js
Comment (tunetheweb) [reply src/browser_action/vitals.js]: I disagree. As per comment above I see these tables as a friendly option than the Performance Entry logging, that could be presented without those. Or could be presented with, but concentrated on in the table and only falling back to the, more difficult-to-read, performance entries if need even more detail. So, with that rational, it seems odd to me to only include one.

I also think the INP table in particular IS useful to have that broken down to it's parts. It is likely we'd be adding that anyway so why not go with it from now.

I agree it can make the console a little noisy, but it's easy to filter by typing the metric name to just show those metrics. The alternative is to have an option per metric but think that's a little overkill.

File: src/browser_action/vitals.js
Comment (tunetheweb) [reply src/browser_action/vitals.js:312]: It's a good question!

Maybe @mmocny can confirm but maybe because it was less mature back then? Or because they are directly referenceable from the Entry itself it wasn't as important to add to the web-vitals build, as that differs slightly from LCP which needs the nav entry for TTFB and then the LCP resource time entry for the other parts so a bit more complicated.

Anyway, as you know Michael's been iterating on this recently so I imagine these sub-parts could change when he adds his "live breakdown" piece after this PR is merged. But still think this is a good first step and good to include for the reasons given above.

But do agree we should add whatever we finally land on to the web-vitals attribution build so other users can grab them easily - at which point we can reference them here too. It wouldn't really simplify the `performance.measure` (which needs both entries anyway) but could simplify the table code slightly (which only needs the duration).

File: src/browser_action/vitals.js
Comment (rviscomi) [reply src/browser_action/vitals.js]: > Done, though used Sentence case as per Optimize LCP guide and usual convention.

Not convinced the guide got that one right to be honest! These metric names are proper nouns and you'll find TTFB in title case everywhere else: https://www.google.com/search?q=site%3Aweb.dev+%22time+to+first+byte%22

For now let's keep it sentence case for parity in this PR and think about updating the guide.

> > Remove LCP
>
> Disagree.

That's fine, but let's write out the full metric name for consistency with the other rows.

File: src/browser_action/vitals.js
Comment (rviscomi) [reply src/browser_action/vitals.js]: Can we compromise on removing the CLS table? Not even sure what to do with this info. We really need the attribution info to take any action on it.

image

File: src/browser_action/vitals.js
Comment (tunetheweb) [reply src/browser_action/vitals.js]: Done.

File: src/browser_action/vitals.js
Comment (mmocny) [reply src/browser_action/vitals.js:327]: We did want to add something to web-vitals attribution, but I decided to wait until we knew better exactly what we want to recommend measuring so as not to break compat eventually.

A long time ago we recommended just doing what we do here in the extension (split a single event into the three components: input-delay, processing-duration, presentation-delay)-- but as Barry says you can do that easily from a single entry.

In practice, processing-duration is better to SUM(all-events) in a single Animation Frame, rather than just looking at a single event.  But when you do that, you end up having to re-define input-delay and presentation-delay to be "delay until first event" and "delay after last event".

Finally, presentation-delay can be split into "rendering-duration" (on main) and "presentation-delay" (off main).

All these things become easier (or possible, even) with Long Animation Frame (LoAF) reporting, and so that's where I've been iterating.


(The direction tooling is going is to "squash" multiple interactions in a single frame down to a single frame latency-- so we'll probably want to evolve this in the extension as well.  I'll try a stab at updating INP user timings after updating INP logs.)

File: src/browser_action/vitals.js
Comment (tunetheweb) [reply src/browser_action/vitals.js]: Actually I've reverted the metric name back to the acronym as that allows me to fit in the rating:

image

**Comment by tunetheweb** [comment]
Already added a few commits ago. Download the latest version!

I went with adding it to the end of the column name (so sometimes elided if your dev tools is small but that's OK as still works for filtering)

image

**Comment by gilbertococchi** [comment]
LGTM, I really liked the Table view for both LCP, CLS and INP. Thanks for reducing the digits a bit on CLS and adding the "[Web Vitals Extension]" label in the Table.

**Comment by mmocny** [comment]
(I'm using the latest, but it was elided because it was too small.)

Actually, playing with console.groupCollapsed api gives me an idea... what if we just merged all the console options, with something like this:

```
console.groupCollapsed('[Web Vitals Extension] INP 200.0', entry);
console.table([{a:1,b:2},{a:3,b:4}]);
console.groupEnd()
```

By default it looks just like a single line:
Screenshot 2023-04-18 at 12 01 07 PM

(It supports showing objects, too, which I did not expect)

But if you expand you see the full table:
Screenshot 2023-04-18 at 12 01 15 PM


This sorta hides table breakdowns by default, so if we hear that folks really want them expanded by default then the Option in settings can turn into:

"Show metric breakdowns by default" or something like that?

**Comment by mmocny** [comment]
Also, I'm not sure I understand what CLS is trying to show?
Screenshot 2023-04-18 at 12 06 19 PM


Why are there several reported shifts with the same score each?

**Comment by mmocny** [comment]
(Barry: Im working on a small patch to try to console.group changes, will send over)

One more thing, once I tried to open the setting page and it just stalled on me.  I noticed I have errors in the extension page and one is related to manifest permissions and the options page:

Screenshot 2023-04-18 at 12 14 20 PM

...wonder if its related to manifest v3, but, I can open the options page most of the time fine, so no idea...

**Comment by mmocny** [comment]
...it just happened again.  I am not sure, but it may depend on the specific tab/page I have opened when I select the options menu item.  I don't see why it should matter, but maybe the handler for the extension contextmenu is trying to access the current page?

**Comment by tunetheweb** [comment]
> Also, I'm not sure I understand what CLS is trying to show?
> 
> Why are there several reported shifts with the same score each?

We only get a shift score per set of shifts. For example if you add a banner image, and it shifts three images down then that's one shift, one shift score, and three elements affected. It would be SOOO much nicer if we got the element causing the shift rather than the victims :-(

Makes a bit more sense when there's two sets of shifts:

image

An alternative would be this, but don't like it myself:

image

WDYT?

**Comment by rviscomi** [comment]
TBH I think we need to pare down the scope of this PR to unblock the P0 changes and focus on #110 and #112. The work on non-LCP tables doesn't feel fully baked yet, so I think it'd be worth speccing that out in an issue first.

**Comment by mmocny** [comment]
That seems fine to leave as-is and to follow-up as needed.

(FWIW even for CLS there is an advantage in just copying things into the table: if the nodes that shift disappear from DOM the attribution disappears from the console.logs... but if you copy it to a table right as the entries are fired, those copies will stay.  It actually helped me a second ago already.)

In other words, +1 to land tables as they are an iterate afterwards.

**Comment by tunetheweb** [comment]
> One more thing, once I tried to open the setting page and it just stalled on me. I noticed I have errors in the extension page and one is related to manifest permissions and the options page:

Extensions cannot access `chrome://` or `chrome-extension://` URLs for security reasons so you see this error when you load that page.

The same thing happens with the production version except you don't get a nice Error screen and instead have to use Inspect Views to see this:

image

**Comment by tunetheweb** [comment]
We could add these URLs to the manifest to avoid this error. But this is more a knock on from #104 rather than this PR.

And it's more obvious as there is an Error screen for local development and you're looking at it a lot.

**Comment by tunetheweb** [comment]
Raised #118 for that issue but, as I say, nothing to do with this PR (nor #104 as it turns out as this was also an issue with Manifests v2).

**Comment by tunetheweb** [comment]
I played around with `console.groupCollapsed` but didn't like it myself. Found it confusing with so many collapsed bits (especially including the object).

I think we want this to be the super simple "easy" path for people wanting a bit more details than the HUD gives, but that don't want to hunt around in the Performance Observer entry that they currently have.

**Comment by tunetheweb** [comment]
I think for live tracing (e.g. for INP) where we show EVERY event doing it collapsed makes sense. But I was seeing that as a separate option for future.

I was seeing these tables as more of a summary view for (ideally only one per metric). Because of the way web-vitals.js works we may see a few more (e.g. an initial LCP and then another LCP) but hopefully not so many that collapsing would be necessary.

But agree people should switch this view off when looking at the future live trace view.

**Comment by tunetheweb** [comment]
Agreed with Michal to merge as is an iterate if needs be.

**Pull request merged**
